File: TypeScript/app/inventory.ts
import { ICustomItem } from './interfaces';

export class Item {
  name: string;
  sellIn: number;
  quality: number;

  constructor(name: string, sellIn: number, quality: number) {
    this.name = name;
    this.sellIn = sellIn;
    this.quality = quality;
  }
}

/* ------------------------------------------ Base Class ------------------------------------------ */
/**
 * abstracts the updating of quality and sellIn values on the Item class dep
 *  */
abstract class CustomItemBase {
  private readonly _item: Item;

  constructor(item: Item) {
    this._item = item;
  }

  get item() {
    return this._item;
  }

  // can be overridden as needed by subclasses like Legendary Item
  protected get maxQuality(): number {
    return 50;
  }

  protected get minQuality(): number {
    return 0;
  }

  // Can be overridden by subclasses like Legendary Item that don't need to be sold
  protected get shouldUpdateSellIn(): boolean {
    return true;
  }

  // default behavior. can be overridden by individual subclasses as needed
  public getRateOfQualityChange() {
    let change = -1;
    if (this.item.sellIn < 0) change = -2;
    return change;
  }

  public update() {
    if (this.shouldUpdateSellIn) {
      this.item.sellIn--;
    }
    const change = this.getRateOfQualityChange();
    let updatedQuality = this.item.quality + change;

    updatedQuality = Math.min(updatedQuality, this.maxQuality); // can't go above max value
    updatedQuality = Math.max(updatedQuality, this.minQuality); // can't go below min value

    this._item.quality = updatedQuality;
  }
}

/* ------------------------------------------ Custom Item Classes ------------------------------------------ */
export class RegularItem extends CustomItemBase implements ICustomItem {}

export class BackstagePasses extends CustomItemBase implements ICustomItem {
  getRateOfQualityChange() {
    const { sellIn, quality } = this.item;
    if (sellIn < 0) return -quality; // worthless after concert date
    let change = 1;
    if (sellIn < 11) change++; // Quality increases by 1 extra point if <11 days to concert (2 total)
    if (sellIn < 6) change++; // Quality increases by 1 extra point if <6 days to concert (3 total)
    return change;
  }
}

/**
 *  overrides maxQuality() and shouldUpdateSellIn() methods
 * */
export class LegendaryItem extends CustomItemBase implements ICustomItem {
  protected get maxQuality(): number {
    return Infinity;
  }

  protected get shouldUpdateSellIn(): boolean {
    return false;
  }

  getRateOfQualityChange() {
    return 0;
  }
}

export class AgedBrie extends CustomItemBase implements ICustomItem {
  getRateOfQualityChange() {
    if (this.item.quality < 50) return 1;
    return 0;
  }
}

export class ConjuredItem extends CustomItemBase implements ICustomItem {
  getRateOfQualityChange() {
    // 	degrades in quality 2x rate of normal item
    // 	get rate of change if this were a normal item
    const regularItem = new RegularItem(this.item);
    const regularChange = regularItem.getRateOfQualityChange();
    return regularChange * 2;
  }
}

File: ruby/inventory.rb
class Item
  attr_accessor :name, :sell_in, :quality

  def initialize(name, sell_in, quality)
    @name = name
    @sell_in = sell_in
    @quality = quality
  end

  def to_s()
    "#{@name}, #{@sell_in}, #{@quality}"
  end
end

# ------------------------------------------ Base Class ------------------------------------------ */
# abstracts the updating of quality and sellIn values in the Item class dependency
class CustomItemBase
  attr_accessor :item

  def initialize(item)
    @item = item
    @max_quality = 50
    @min_quality = 0
  end

  def legendary?
    false
  end

  private

  attr_reader :max_quality, :min_quality

  def update_quality(change)
    updated_quality = @item.quality + change
    unless legendary?
      updated_quality = [updated_quality, max_quality].min
      updated_quality = [updated_quality, min_quality].max
    end

    @item.quality = updated_quality
  end

  def update_sell_in
    @item.sell_in -= 1 unless legendary?
  end

  public

  def update(change)
    update_sell_in
    update_quality(change)
  end
end


# ------------------------------------------ Custom Item Classes ------------------------------------------ */
# getRateOfQualityChange() defines item-specific logic
# isLegendary (optional) bypasses max quality value and sellIn change
class RegularItem < CustomItemBase
  def rate_of_quality_change
    change = -1
    change = -2 if item.sell_in.negative?
    change
  end
end

class BackstagePasses < CustomItemBase
  def rate_of_quality_change
    sell_in = item.sell_in
    quality = item.quality

    return -quality if sell_in.negative?

    change = 1
    change += 1 if sell_in < 11
    change += 1 if sell_in < 6

    change
  end
end

class LegendaryItem < CustomItemBase
  def rate_of_quality_change
    0
  end

  def legendary?
    true
  end
end

class AgedBrie < CustomItemBase
  def rate_of_quality_change
    1
  end
end

class ConjuredItem < CustomItemBase
  def rate_of_quality_change
    # 	degrades in quality 2x rate of normal item
    # 	get rate of change if this were a normal item
    regular_item = RegularItem.new(item)
    regular_change = regular_item.rate_of_quality_change
    regular_change * 2
  end
end


